ext/spl/php_spl.c

```diff
@@ -508,86 +508,46 @@ PHP_FUNCTION(spl_autoload_call)
 PHP_FUNCTION(spl_autoload_register)
 {
 	zend_string *func_name;
-	char *error = NULL;
 	zend_string *lc_name;
-	zval *zcallable = NULL;
 	zend_bool do_throw = 1;
 	zend_bool prepend  = 0;
 	zend_function *spl_func_ptr;
 	autoload_func_info alfi;
 	zend_object *obj_ptr;
+	zend_fcall_info fci = {0};
 	zend_fcall_info_cache fcc;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|z!bb", &zcallable, &do_throw, &prepend) == FAILURE) {
-		RETURN_THROWS();
+	ZEND_PARSE_PARAMETERS_START(0, 3)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_FUNC_EX(fci, fcc, /* check null */ 1, /* separate */ 0)
+		Z_PARAM_BOOL(do_throw)
+		Z_PARAM_BOOL(prepend)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (!do_throw) {
```

Comment:
Do you know why this argument exists in the first place (e.g. from commit history)? It's very weird, but I feel like there must be some reason it was added...

Reply:
I'll look into it, because I'm quite baffled by it too.

Reply:
Seems like it was added 12 years ago without much explanation: https://github.com/php/php-src/commit/0cfdd9a7e078c5317e328df5444c5c5a2a11e5df

Reply:
Reason for change introduction: https://bugs.php.net/bug.php?id=42823
Thanks to @IMSoP for finding the bug report.

Reply:
composer defaults to true: https://github.com/composer/composer/commit/c80cb76b9b5082ecc3e5b53b1050f76bb27b127b

Docs: https://getcomposer.org/doc/06-config.md#prepend-autoloader

Seems it's pretty critical, except if we make the prepend behaviour the default and deprecate the last 2 arguments.
Or maybe a better idea is to introduce a new function ``autoload_register(callable $callable, bool $prepend)`` but that means changing all SPL related autoload functions which doesn't seem really feasible.

Reply:
Hmm, isn't this about `do_throw`? That has been introduced with a5c37f37b6ea73c0087f930582c6ab66b75c8d56 ("Add ability to fail silently").

Reply:
Err, sorry, I think was referring to the `do_throw` argument here, that gets ignored by this patch. I can see the usefulness of `prepend`, but not so much of `do_throw`.

Reply:
Ah my mistake, I would say the consistent TypeError RFC would back this change, but it does make the signature awkward